unison-runtime/src/Unison/Runtime/Stack.hs

```diff
@@ -284,13 +363,118 @@ frameDataSize = go 0
     go sz (Push f a _ _ _ k) =
       go (sz + f + a) k
 
-pattern DataC :: Reference -> Word64 -> SegList -> Closure
+pattern DataC :: Reference -> PackedTag -> SegList -> Closure
 pattern DataC rf ct segs <-
   (splitData -> Just (rf, ct, segs))
   where
     DataC rf ct segs = formData rf ct segs
 
-type SegList = [Either Int Closure]
+-- | An unboxed value with an accompanying tag indicating its type.
+data TypedUnboxed = TypedUnboxed {getTUInt :: !Int, getTUTag :: !PackedTag}
+  deriving (Show, Eq)
```

Comment:
So, a question I have is: is there a reason to use this type instead of `Val`?

You might say it's smaller, but storing just the unboxed type tag means if we later write it back to the stack we'll have to allocate a new `Closure` indicating the type, which could have just been shared with wherever we got this (and/or the top-level ones you define elsewhere). So it doesn't seem like it actually saves anything.

I guess it's representing fewer incorrect states. But I'm not sure how big of a deal that is.

Reply:
Yeah I think you're right, I added this type before realizing I'd need `Val`, and haven't circled back to simplify yet but I think that's a good idea.

Part of it is that the current Unboxed-as-Boxed setup is still a bit screwed up and there are two different representations for unboxed values there that I need to sort out.

On trunk, we put them inside a `DataU1` with the ref and tag of the unboxed type, but I think now that Val is propagated everywhere  we just don't put unboxed values in closures at all (unless a field of a  boxed type is unboxed of course)

So yeah, I'll try removing this and see what it looks like  👍🏼